plugins-root/check_icmp.c

```diff
@@ -1160,31 +1323,59 @@ add_target_ip(char *arg, struct in_addr *in)
 static int
 add_target(char *arg)
 {
-	int i;
-	struct hostent *he;
-	struct in_addr *in, ip;
+	int error, result;
+	struct sockaddr_storage ip;
+	struct addrinfo hints, *res, *p;
+	struct sockaddr_in *sin;
+	struct sockaddr_in6 *sin6;
+
+	switch (address_family) {
+	case -1:
+		// -4 and -6 are not specified on cmdline
+		address_family = AF_INET;
+		sin = (struct sockaddr_in *)&ip;
+		result = inet_pton(address_family, arg, &sin->sin_addr);
+#ifdef USE_IPV6
+		if( result != 1 ){
+			address_family = AF_INET6;
+			sin6 = (struct sockaddr_in6 *)&ip;
+			result = inet_pton(address_family, arg, &sin6->sin6_addr);
+		}
+#endif
+		break;
+	case AF_INET:
+		sin = (struct sockaddr_in *)&ip;
+		result = inet_pton(address_family, arg, &sin->sin_addr);
+		break;
+	case AF_INET6:
+		sin6 = (struct sockaddr_in6 *)&ip;
+		result = inet_pton(address_family, arg, &sin6->sin6_addr);
+		break;
+	default: crash("Address family not supported");
+	}
 
 	/* don't resolve if we don't have to */
-	if((ip.s_addr = inet_addr(arg)) != INADDR_NONE) {
+	if(result == 1) {
 		/* don't add all ip's if we were given a specific one */
 		return add_target_ip(arg, &ip);
-		/* he = gethostbyaddr((char *)in, sizeof(struct in_addr), AF_INET); */
-		/* if(!he) return add_target_ip(arg, in); */
 	}
 	else {
 		errno = 0;
-		he = gethostbyname(arg);
-		if(!he) {
+		memset(&hints, 0, sizeof(hints));
+		hints.ai_family = AF_UNSPEC;
```

Comment:
AF_UNSPEC should only be used if `address_family` is -1. Otherwise we will set address_familiy to v6 even if -4 is specified on the command line.
ex.: `./check_icmp -H labs.consol.de -4 -v` uses the ipv6 address.

Reply:
the commandline interface seams not self explained.

one time you requested 
   ` ./check_icmp -4 -H ipv4host -6 -H ipv6host`
to work. the position of -4 and -6 is important.

in case of 
  `./check_icmp -H localhost -4`
you expected `-4` should change the name resolution of `localhost`. but if you want position dependend parameter handling, this call is not valid.

just a theoretical question. how about:
`./check_icmp -H localhost -4 -H ipv4host -6 -H ipv6host`
would you expect an IPv4 or IPv6 check for localhost? If localhost has a dual stack, i'm expecting localhost and  ipv6host are checked with IPv6 and ipv4host with IPv4. localhost should be checked by IPv6 because the OS would also prefer IPv6 over IPv4.

i don't have such complex use cases. i only need a working auto detection for any IPv4 and IPv6 address without specifying -4 or -6. if an address is specified, there is no option for interpretation. so -4 and -6 are redundant.
`./check_icmp ::1`
`./check_icmp 127.0.0.1`

Reply:
right now, it is not possible to mix ipv4 and ipv6 addresses. Maybe we should emit an error if somebody tries too.